src/Pool.sol

```diff
@@ -563,110 +585,141 @@ abstract contract Pool {
-    function _assertCurrReceivers(Sender memory sender, Receiver[] calldata currReceivers)
-        internal
-        pure
-    {
-        require(hashReceivers(currReceivers) == sender.receiversHash, "Invalid current receivers");
+    /// @notice Asserts that the sender state is the currently used one.
+    /// @param senderId The id of the sender
+    /// @param lastUpdate The timestamp of the last update of the sender.
+    /// If this is the first update of the sender, pass zero.
+    /// @param lastBalance The balance after the last update of the sender.
+    /// If this is the first update of the sender, pass zero.
+    /// @param currReceivers The list of receivers set in the last update of the sender.
+    /// If this is the first update of the sender, pass an empty array.
+    function _assertSenderState(
+        SenderId memory senderId,
+        uint64 lastUpdate,
+        uint128 lastBalance,
+        Receiver[] calldata currReceivers
+    ) internal view {
+        bytes32 expectedHash;
+        if (senderId.isSubSender) {
+            expectedHash = subSenderStateHash[senderId.senderAddr][senderId.subSenderId];
+        } else {
+            expectedHash = senderStateHash[senderId.senderAddr];
+        }
+        bytes32 actualHash = hashSenderState(lastUpdate, lastBalance, currReceivers);
+        require(actualHash == expectedHash, "Invalid provided sender state");
     }
 
-    /// @notice Calculates the hash of the list of receivers.
-    /// @param receivers The list of the receivers.
+    /// @notice Stores the hash of the updated sender state to be used in `_assertSenderState`.
+    /// @param senderId The id of the sender
+    /// @param newBalance The new sender balance.
+    /// @param newReceivers The new list of the sender's receivers.
     /// Must be sorted by the receivers' addresses, deduplicated and without 0 amtPerSecs.
-    /// @return receiversHash The hash of the list of receivers.
-    function hashReceivers(Receiver[] calldata receivers)
-        public
-        pure
-        returns (bytes32 receiversHash)
-    {
-        if (receivers.length == 0) return bytes32(0);
-        return keccak256(abi.encode(receivers));
-    }
-
-    /// @notice Returns the sender's receivers list hash.
-    /// @param senderAddr The address of the sender
-    /// @return receiversHash The receivers list hash.
-    function getReceiversHash(address senderAddr) public view returns (bytes32 receiversHash) {
-        return senders[senderAddr].receiversHash;
+    function _storeSenderState(
+        SenderId memory senderId,
+        uint128 newBalance,
+        Receiver[] calldata newReceivers
+    ) internal {
+        bytes32 stateHash = hashSenderState(_currTimestamp(), newBalance, newReceivers);
+        if (senderId.isSubSender) {
+            subSenderStateHash[senderId.senderAddr][senderId.subSenderId] = stateHash;
+        } else {
+            senderStateHash[senderId.senderAddr] = stateHash;
+        }
```

Comment:
I think we should have a seperate `UpdateSenderHash`event. The `_currTimestamp`/ `lastUpdate` information is currently not part of any event.  Implict because available because of the block.timestamp.

Reply:
However, it might be nice to have the data and the hash in the same event.

Reply:
`block.timestamp` of emission is a part of EVERY event :smile: `SenderUpdated` event contains all the data needed to both recreate the hash and call `updateSender`, IMO putting hash into the event would be just a waste of gas, if one needs the hash off-chain, they can easily calculate it for free.

Reply:
Yeah, I know it is available. I think if we provide a low-level javascript library everything should be fine.